### DISCUSSION TITLE
Extending LightEntities in a generic way

### DISCUSSION THREAD

**Discussion by Breina** [home-assistant/architecture#800] [Entity Models] [answered]
# Problem definition

I’m a developer of the DMX integration, a protocol which is generally used for party lights. People have been using it to control lights in their home.

Some lights however, don’t fit into the [light color modes provided](https://github.com/home-assistant/core/pull/47720). Some of these lights have channels like [Amber and UV](https://github.com/jnimmo/hass-dmx/issues/52), [others have parameters like pan, tilt and strobing frequency](https://github.com/jnimmo/hass-dmx/issues/62).

# Just extending the current system

The current supported color modes are as follows:

    ColorMode.ONOFF,
    ColorMode.BRIGHTNESS,
    ColorMode.COLOR_TEMP,
    ColorMode.HS,
    ColorMode.XY,
    ColorMode.RGB,
    ColorMode.RGBW,
    ColorMode.RGBWW,
    ColorMode.WHITE,

Just adding more light color modes won't help, as there would be too many permutations of all desired parameters. I feel like we are running into the limitations of the LightEntity architecture currently.

# Proposal

We could move away from predefined light color modes and replace it with a more generic system to define light channels. For example:

    lightEntity.addChannel(ChannelType.COLOR_TEMP, min=150, max=450)
    lightEntity.addChannel(ChannelType.RGB)
    lightEntity.addChannel(ChannelType.CUSTOM, name="Strobing frequency", min=0, max=100)

# Compatibility

It might not be worth breaking hundreds of integrations that currently depend on the ColorMode system. Perhaps there's a compromise needing to be made here?

Mainly looking for feedback currently, other proposals definitely welcome!

**Comment by Jc2k** [comment]
Modern Home Assistant tries to decompose things into entities that are related by the device registry. Strobing frequency, pan and tilt in particular feel like they should be handled like that.

**Reply by Breina** [reply]
Thanks for the quick reply! Yeah I guess that's fair for those parameters. I can have a device with one LightEntity and multiple NumberEntities.

Might be sufficient to add a couple of permutations for UV in that case. Counting amber as warm white, it could suffice just adding a permutation on color modes for UV?

**Comment by frenck** [comment, marked as answer]
From the example:

> `lightEntity.addChannel(ChannelType.CUSTOM, name="Strobing frequency", min=0, max=100)`

The problem with these things it that the UI will need to be included, which is hard on random data and parameters, this includes things like automations. Home Assistant doesn't want to support every possible permutation/capability a device has; that is not our goal.

We can add support for a new color mode, but that requires multiple integrations that can leverage such a mode.

Our goal is to provide a common interface that fits most of the devices out there. Unfortunately, outliers, like this case, will not fit our model, which is fine.

We do offer enough controls to still make these work. For example, in this case, a simple light entity with on/off capabilities and use entities like `select` & `number` to provide controls over the strobe and strobe speed. Another approach could be adding Strobe as an effect and combining that with a number entity to control the strobe speed.

../Frenck

**Reply by Breina** [reply]
Alright, that seems completely fair, thanks for your reply!

>We can add support for a new color mode, but that requires multiple integrations that can leverage such a mode.

I will be looking for lobbyists! :)
